File: packages/hpack-html-py/hpack_html/format.py
"""Format constants for the .hpack binary format."""

MAGIC = b"\x89\x48\x50\x4B"
MAGIC_SIZE = 4
VERSION = 0x01

FLAG_MINIFIED = 1 << 0
FLAG_CHECKSUM = 1 << 1
FLAG_COMPRESSION_MASK = 0x0C
FLAG_COMPRESSION_SHIFT = 2

COMPRESSION_DEFLATE = 0
COMPRESSION_GZIP = 1

FIELD_URL = 0x01
FIELD_ETAG = 0x02
FIELD_SIGNATURE = 0x03
FIELD_CONTENT_TYPE = 0x04
FIELD_TIMESTAMP = 0x05
FIELD_ENCODING = 0x06
FIELD_CUSTOM_START = 0x10

CRC32_SIZE = 4
MIN_PACKET_SIZE = MAGIC_SIZE + 1 + 1 + 1  # magic + version + flags + header_len


def encode_flags(minified: bool, checksum: bool, compression: int) -> int:
    flags = 0
    if minified:
        flags |= FLAG_MINIFIED
    if checksum:
        flags |= FLAG_CHECKSUM
    flags |= (compression & 0x03) << FLAG_COMPRESSION_SHIFT
    return flags


def decode_flags(flags: int) -> tuple[bool, bool, int]:
    """Returns (minified, checksum, compression)."""
    return (
        bool(flags & FLAG_MINIFIED),
        bool(flags & FLAG_CHECKSUM),
        (flags & FLAG_COMPRESSION_MASK) >> FLAG_COMPRESSION_SHIFT,
    )

File: packages/hpack-html-go/format.go
package hpack

// Magic bytes: 0x89 "HPK"
var Magic = []byte{0x89, 0x48, 0x50, 0x4B}

const (
	MagicSize   = 4
	Version     = 0x01
	VersionSize = 1
	FlagsSize   = 1

	FlagMinified        = 1 << 0
	FlagChecksum        = 1 << 1
	FlagCompressionMask = 0x0C
	FlagCompressionShift = 2

	CompressionDeflate = 0
	CompressionGzip    = 1

	FieldURL         = 0x01
	FieldETag        = 0x02
	FieldSignature   = 0x03
	FieldContentType = 0x04
	FieldTimestamp   = 0x05
	FieldEncoding    = 0x06
	FieldCustomStart = 0x10

	CRC32Size     = 4
	MinPacketSize = MagicSize + VersionSize + FlagsSize + 1
)

// EncodeFlags encodes options into a flags byte.
func EncodeFlags(minified, checksum bool, compression int) byte {
	var flags byte
	if minified {
		flags |= FlagMinified
	}
	if checksum {
		flags |= FlagChecksum
	}
	flags |= byte((compression & 0x03) << FlagCompressionShift)
	return flags
}

// DecodedFlags holds parsed flag values.
type DecodedFlags struct {
	Minified    bool
	Checksum    bool
	Compression int
}

// DecodeFlags parses a flags byte.
func DecodeFlags(flags byte) DecodedFlags {
	return DecodedFlags{
		Minified:    flags&FlagMinified != 0,
		Checksum:    flags&FlagChecksum != 0,
		Compression: int((flags & FlagCompressionMask) >> FlagCompressionShift),
	}
}
